### ISSUE TITLE
How to release keras compiled model

### ISSUE THREAD

**Issue by fsx950223** [tensorflow/tensorflow#31026]
<em>Please make sure that this is a bug. As per our [GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md), we only address code/doc bugs, performance issues, feature requests and build/installation issues on GitHub. tag:bug_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow):No
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04):Ubuntu 18.04
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device:
- TensorFlow installed from (source or binary):
- TensorFlow version (use command below):1.14.0
- Python version:3.6.7
- Bazel version (if compiling from source):
- GCC/Compiler version (if compiling from source):
- CUDA/cuDNN version:
- GPU model and memory:

You can collect some of this information using our environment capture
[script](https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh)
You can also obtain the TensorFlow version with: 1. TF 1.0: `python -c "import
tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"` 2. TF 2.0: `python -c
"import tensorflow as tf; print(tf.version.GIT_VERSION, tf.version.VERSION)"`

**Describe the current behavior**
It seems that tensorflow will increase memory usage when I compiled keras model serveral time.
There is some way to release unused compiled model 
**Describe the expected behavior**
Release compiled model when I don't need it anymore
**Code to reproduce the issue**
Provide a reproducible test case that is the bare minimum necessary to generate the problem.
``` python
model.compile()
"""freeze training"""

model.compile()
"""fine-tuning training"""
```
Replicate code: https://colab.research.google.com/drive/1GafCUt6BXjKY0LA3lc8fgGhDPwZ6Jgs3

**Other info / logs**
Include any logs or source code that would be helpful to diagnose the problem. If including tracebacks, please include the full traceback. Large logs and files should be attached.

**Comment by gadagashwini** [comment]
@fsx950223 Will it be possible to share the minimal code which replicates the reported issue. Thanks!

**Comment by fsx950223** [comment]
> @fsx950223 Will it be possible to share the minimal code which replicates the reported issue. Thanks!

I provided code above

**Comment by jvishnuvardhan** [comment]
@fsx950223 Can you try `tf.reset_default_graph()` after importing tensorflow. I have noticed that memory was not increasing after adding `tf.reset_default_graph()`. Please check the [gist here](https://colab.sandbox.google.com/gist/jvishnuvardhan/fa7eb1e000b42ac27f8a56a286d29f84/tf_31026_memory.ipynb). Thanks!

**Comment by fsx950223** [comment]
> @fsx950223 Can you try `tf.reset_default_graph()` after importing tensorflow. I have noticed that memory was not increasing after adding `tf.reset_default_graph()`. Please check the [gist here](https://colab.sandbox.google.com/gist/jvishnuvardhan/fa7eb1e000b42ac27f8a56a286d29f84/tf_31026_memory.ipynb). Thanks!

I ran the gist and memory was still increasing after keras model compiled and I can't release it

**Comment by jvishnuvardhan** [comment]
@fsx950223 Could throw little more details on the issue? How (sequence) did you ran the code. 

The way i ran the model the gist is, first reset_graph, ran create model (0.6 GB of ram), then compiled, then ran training and found 0.85 GB was using.

Then ran entire gist (reset_graph, create_model, compile, training) again and it was taking only 0.85 GB of ram. 

Please share the sequence and any other details. Thanks!

**Comment by fsx950223** [comment]
``` python
for i in range(10):
  model.compile(optimizer='adam',
                loss='sparse_categorical_crossentropy',
                metrics=['accuracy'])
  #model.release() # Release memory
  print(psutil.Process(os.getpid()).memory_info().rss) # You can see memory increase 
```
The Problem is that memory will increase after I compiled model. I need some method to release compiled model and release memory.

**Comment by fsx950223** [comment]
> @fsx950223 Could throw little more details on the issue? How (sequence) did you ran the code.
> 
> The way i ran the model the gist is, first reset_graph, ran create model (0.6 GB of ram), then compiled, then ran training and found 0.85 GB was using.
> 
> Then ran entire gist (reset_graph, create_model, compile, training) again and it was taking only 0.85 GB of ram.
> 
> Please share the sequence and any other details. Thanks!

for i in range(10):
  model.compile(optimizer='adam',
                loss='sparse_categorical_crossentropy',
                metrics=['accuracy'])
  #model.release() # Release memory
  print(psutil.Process(os.getpid()).memory_info().rss) # You can see memory increase 
The Problem is that memory will increase after I compiled model. I need some method to release compiled model and release memory.

**Comment by jvishnuvardhan** [comment]
@fsx950223 I agree that memory is increasing. What is the reason behind compiling multiple times with same loss and optimizer? Thanks!

**Comment by fsx950223** [comment]
> @fsx950223 I agree that memory is increasing. What is the reason behind compiling multiple times with same loss and optimizer? Thanks!

Maybe I want to change trainable layers or distribute strategy that I need to compile it again. Although I can solve the problem by run python script several times, it will cost more time to train the model.

**Comment by jvishnuvardhan** [comment]
@fsx950223 When you change anything related to what you listed or anything related model, optimizer, and loss function, model need to be compiled and need to train the model again. If I am correct, it was adding only 1MB when you recompile the given example. Thanks!

**Comment by fsx950223** [comment]
It's a example. In fact, it increased 8g memory in my project

Vishnuvardhan Janapati <notifications@github.com<mailto:notifications@github.com>> 于 2019年8月23日周五 上午1:56写道：

@fsx950223<https://github.com/fsx950223> When you change anything related to what you listed or anything related model, optimizer, and loss function, model need to be compiled and need to train the model again. If I am correct, it was adding only 1MB when you recompile the given example. Thanks!

**Comment by robieta** [comment]
@fsx950223 Can you provide an example with the more pathological behavior? As @jvishnuvardhan said, a small increase in memory consumption on recompile is expected.

**Comment by fsx950223** [comment]
> @fsx950223 Can you provide an example with the more pathological behavior? As @jvishnuvardhan said, a small increase in memory consumption on recompile is expected.

I have updated notebook. I just tested it with a small dataset, and large dataset will significantly improve memory usage.

**Comment by jvishnuvardhan** [comment]
I am closing this issue as the increase is very minimal. Please feel free to reopen with more concrete example and a standalone code to reproduce the issue. Thanks!
